src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/XSharedTransformer.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.AstNodes;
+using XamlX;
+using XamlX.Ast;
+using XamlX.Emit;
+using XamlX.IL;
+using XamlX.Transform;
+using XamlX.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers;
+
+class XSharedTransformer : IXamlAstTransformer
+{
+    public IXamlAstNode Transform(AstTransformationContext context, IXamlAstNode node)
+    {
+        var types = context.GetAvaloniaTypes();
+        switch (node)
+        {
+            case XamlPropertyAssignmentNode { Values.Count: 2 } pa when
+                    pa.Property.Name == "Content"
+                    && pa.Property.DeclaringType == types.ResourceDictionary
+                    && pa.Values[1] is XamlAstConstructableObjectNode co
+                    && TryGetSharedDirective(co, out var directive)
+                    :
+                {
+                    co.Children.Remove(directive.Node);
+                    if (!directive.Value)
+                    {
+                        pa.Values[1] = new XamlNotSharedDeferredContentNode(pa.Values[1]
+                            , types.NotSharedDeferredContentExecutorCustomization
+                            , types.XamlIlTypes.Object
+                            , context.Configuration);
+                        pa.PossibleSetters = new List<IXamlPropertySetter>
+                        {
+                            new XamlDirectCallPropertySetter(types.ResourceDictionaryNotSharedAdd),
+                        };
+                    }
+                }
+                break;
+            case XamlPropertyAssignmentNode { Values.Count: 2 } pa when
+                    pa.Property.Name == "Resources"
+                    && pa.Property.Getter?.ReturnType.Equals(types.IResourceDictionary) == true
+                    && pa.Values[1] is XamlAstConstructableObjectNode co
+                    && TryGetSharedDirective(co, out var directive)
+                    :
+                {
+                    co.Children.Remove(directive.Node);
+                    if (!directive.Value)
+                    {
+                        pa.Values[1] = new XamlNotSharedDeferredContentNode(pa.Values[1]
+                            , types.NotSharedDeferredContentExecutorCustomization
+                            , types.XamlIlTypes.Object
+                            , context.Configuration);
+                        pa.PossibleSetters = new List<IXamlPropertySetter>
+                        {
+                            new AdderSetter(pa.Property.Getter, types.ResourceDictionaryNotSharedAdd),
+                        };
+
+                    }
+                }
+                break;
+            default:
+                break;
+        }
+
+        return node;
+
+        bool TryGetSharedDirective(XamlAstConstructableObjectNode co, out (IXamlAstNode Node, bool Value) directive)
```

Comment:
Nullability: on `false` return, `directive.Node` will be `null`, which isn't expected by the non-nullable `IXamlAstNode`.

Prefer two parameters `[NotNullWhen(true)] out IXamlAstNode? node` and `out bool value` to better represent the return values.